### ISSUE TITLE
Ability to request multiple simultaneous sections of text

### ISSUE THREAD

**Issue by monotasker** [distributed-text-services/specifications#138]
I'm trying to design an API that will conform to the DTS spec. It looks, though, like the documents endpoint is designed to allow only one ref request at a time. To minimize the number of requests for clients, we would like to be able to allow one request that specifies several ref values and returns the text for all of them in a single response. Is this possible and I've just missed it? If not, is there any interest in allowing for multiple refs in a request?

**Comment by PonteIneptique** [comment]
Hi @monotasker ! Welcome !
It was something we talked about during the development of the specs, and we put on the side at the time, I think mostly because we wanted to get the main frame correct.

I do think that is something we need to talk about again. 

Here are my concerns (mostly technical ones) : 
1. with the current specs, how would the `Link` section of a document request behave in case multiple sections are requested ?
2. with the current specs, how would we deal with range request ? I see easily asking for `?ref=1&ref=3&ref=5` but it feels like `?start=1&end=2&start=4&end=5` could be quite quickly buggy. I can completely be wrong.
3. is the overhead induced by splitting requests important for the client / server ? I think this is also a big part of moving forward this kind of request. To me, it feels mostly like HTTP overhead but I could be wrong.
4. if we'd do so, we would need (but the investment is not huge) to insert in `<dts:fragment/>` attributes that helps identifying each passage. Not a huge cost at all, just pointing it here.

Overall I feel like the proposal is good but 1 and 2 are worrying me, while 3 can easily help us make the decision. I would really like to have input of the group here :)

**Comment by balmas** [comment]
I will add one to @PonteIneptique 's list:

5. what would error responses look like for this? What if the response is mixed success and error?

I do think this is going to be a common enough use case that we need to address it, but it might be challenging to find answers to the questions that will be general enough to suit most needs.

**Comment by balmas** [comment]
Per the 2018-10-04 DTS Meeting, discussion of this item will be on the agenda for next month's meeting (1 November 2018). Interested parties to please review and provide opinions prior to the meeting.

**Comment by monotasker** [comment]
Thanks for considering this. Here are a couple of my thoughts.

### Re. issue 2 above (requesting multiple ranges)
It seems to me like the simplest approach would be to allow for requesting an array of start/end pairs using standard js array/object syntax. So one might allow this:

`?ref=[{start:2, end:4}, {start:5, end:12}, 36]`

Notice that the third value here is just a straight ref id. This approach would allow easy mixing of requests for ranges and for single textual units.

### Re. issue 1 above (handling the Link response header)
A response can include multipe Link headers. So returning separate link information for each requested range isn't a problem. The only trick would be identifying which Link header is associated with which range. Could we use either the "target" or the "title" attribute of each link to identify the ref for which the link is intended? E.g. for the multi request above, the Link headers might be:
~~~
Link: </dts/api/documents/?id=bgu;11;2029&ref=1>; rel=”prev”; title="{start:2, end:4}",
        </dts/api/documents/?id=bgu;11;2029&ref=5>; rel=”next”; title="{start:2, end:4}", 
        </dts/api/documents/?id=bgu;11;2029&ref=37>; rel=”last”, 
        </dts/api/navigation/?id=bgu;11;2029>; rel=”contents”, 
        </dts/api/collection/?id=bgu;11;2029>; rel=”collection”
Link: </dts/api/documents/?id=bgu;11;2029&ref=4>; rel=”prev”; title="{start:5, end:12}", 
        </dts/api/documents/?id=bgu;11;2029&ref=13>; rel=”next”; title="{start:5, end:12}", 
        </dts/api/documents/?id=bgu;11;2029&ref=37>; rel=”last”, 
        </dts/api/navigation/?id=bgu;11;2029>; rel=”contents”, 
        </dts/api/collection/?id=bgu;11;2029>; rel=”collection”
Link: </dts/api/documents/?id=bgu;11;2029&ref=35>; rel=”prev”; title="36", 
        </dts/api/documents/?id=bgu;11;2029&ref=37>; rel=”next”; title="36", 
        </dts/api/documents/?id=bgu;11;2029&ref=37>; rel=”last”, 
        </dts/api/navigation/?id=bgu;11;2029>; rel=”contents”, 
        </dts/api/collection/?id=bgu;11;2029>; rel=”collection” 
~~~
Semantically I think it would make sense to include this identifying information on the "prev" and "next" links, since those are the only two that change depending on the requested range.

This could probably be compressed into one Link header like this:

Link: </dts/api/documents/?id=bgu;11;2029&ref=1>; rel=”prev”; title="{start:2, end:4}",
        </dts/api/documents/?id=bgu;11;2029&ref=5>; rel=”next”; title="{start:2, end:4}", 
        </dts/api/documents/?id=bgu;11;2029&ref=4>; rel=”prev”; title="{start:5, end:12}", 
        </dts/api/documents/?id=bgu;11;2029&ref=13>; rel=”next”; title="{start:5, end:12}", 
        </dts/api/documents/?id=bgu;11;2029&ref=35>; rel=”prev”; title="36", 
        </dts/api/documents/?id=bgu;11;2029&ref=37>; rel=”next”; title="36", 
        </dts/api/documents/?id=bgu;11;2029&ref=37>; rel=”last”, 
        </dts/api/navigation/?id=bgu;11;2029>; rel=”contents”, 
        </dts/api/collection/?id=bgu;11;2029>; rel=”collection” 

The only disadvantage of this approach would be that if at some point you wanted to allow multiple requests *for different documents* the multiple Link headers might make that semantically easier. If you're not concerned about that, then maybe this short format would work.

**Comment by PonteIneptique** [comment]
I'd actually like to remind why we postponed this feature early : we thought the HTTP overhead would be okay, while delivery could be quite complicated to implement.

**Comment by PonteIneptique** [comment]
Note: This last comment is just a memory of a really old meeting. This does not in any case condemn this issue at all.

**Comment by monotasker** [comment]
The advantage here is a major one for building clients. I edit the Online Critical Pseudepigrapha, where our interface allows users to view several versions of a document simultaneously, in parallel. It takes a noticeable amount of time to send a request for one text section to the server and present the result. But if I have to send a separate request for each version being displayed, then the time a user spends waiting for the result of navigation can in some cases be multiplied by the number of requests. If I'm viewing 4 versions in parallel, that's the difference between waiting .5 seconds after I click "next" and waiting 2 seconds. A .5 second delay feels acceptable in a UI. A 2 second delay begins to feel frustrating. 

To be clear, the advantage I have in mind here isn't processing time, either on the server or in the client. The advantage lies in avoiding the time waiting for multiple requests to be returned if the server ends up processing them sequentially. That's why so many commercial APIs allow batch requests.

**Comment by PonteIneptique** [comment]
For the use case here, I think our original argument would be that the use *async* would lessen the wait potentially ?

Though, I completely understand your request, and I think we need to evaluate it, even if it take a little time to get to a point where we are ok with it. Would you be available for next call ?

**Comment by monotasker** [comment]
@PonteIneptique, since I'm new here, I'm not quite sure what "next call" refers to. But I'm happy to talk in another context if that's useful.

**Comment by PonteIneptique** [comment]
Oups, sorry, we really to work a bit on how to welcome people :) We have a call on November 1 2018 (4PM CET), that might be nice to meet you and start discussing this major feature in context ?

**Comment by jonathanrobie** [comment]
I would rather do a more general query feature that allows keyword/value pairs to be queried.  Any number of results can be returned.

One approach is to have a string-based little language that fits in a URI and use it with GET.  That's nicely pure.  A less pure way to do it is to POST a query document to a query endpoint.

**Comment by monotasker** [comment]
I can see the value of @jonathanrobie 's suggestion of a query feature. On the other hand, I can also see how waiting to address this use case until we design a full query system might not be workable. As a compromise, maybe we could introduce a new parameter (as in @hcayless 's suggestion) that can take a query string. In designing the structure for that query string we can be thinking in terms of generalizing the syntax down the road in a more complete query system. So we could get the multiple request use-case addressed more quickly (and arguably in a more user-friendly way) and still have that work lay the foundation for a future query system.

By way of syntax, we could model it on an array of js objects, each of which has the same keys as the URL parameters in a single request. For example:

```js
[
  {id: mydocument,
   start:3,
   end:7},
  {id: mynextdocument,
   ref:"4:36"}
]
```
Or, compressed as a string in the url parameter:

`[{id: "mydocument",start:"3",end:"7"},{id: "mynextdocument",ref:"4:36"}]`

Note that I've included the use of quotation marks around the values so that a colon can be used as a delimiter in ref values.

**Comment by PonteIneptique** [comment]
I would vote for this solution, if it were actually JSON instead of JS (for
ease of parsing in languages...)

Le mar. 6 nov. 2018 à 18:02, Ian W. Scott <notifications@github.com> a
écrit :

**Comment by monotasker** [comment]
@PonteIneptique , I believe my example *is* an array of JSON objects, n'est pas? JSON is just javascript objects, right? Since we want multiple, we put them in a JS array. Or am I missing something?

**Comment by PonteIneptique** [comment]
Keys in javascript do not need to be encapsulated in quotes, unlike in json
AFAIK. This is a big different for parsers

Le mar. 6 nov. 2018 à 9:05 PM, Ian W. Scott <notifications@github.com> a
écrit :

**Comment by monotasker** [comment]
Right. That's my mistake. I've been working in Python more than JS for a while.

**Comment by jonathanrobie** [comment]
I wish there were a way to let a user provide a list of query strings in a single request.  I don't think there is, but I could be wrong.  I don't like making the client do the parsing here.

**Comment by jonathanrobie** [comment]
I suspect I would like to use the same convention to return several things for:

1. Queries
2. Multiple fragments
3. Multiple specified documents

Is that right?  If so, let's make sure the return conventions we use here will work well for other similar settings.

**Comment by monotasker** [comment]
My thinking here was that a JSON string is actually the only de facto standard for encoding multiple sets of request parameters like that. You're right that the server can't parse it. But it's at least a format that any client designer will be familiar with. And I think it's a flexible enough format to accommodate all the kinds of returns you're talking about.

**Comment by jonathanrobie** [comment]
Another possibility is to have a query language that has only:

- parentheses
- and, or
- keyword, value pairs

If need be, we could restrict the set of keywords that are allowed to those needed in this use case.

```
?q="(id='urn:xxxx' and start='1.1.1' and end='1.2.1') or (id='urn:yyy' and start='2.1.1' and end='2.2.1')"
```

**Comment by PonteIneptique** [comment]
Would actually this work : 

`?id[0]=urn:xxx&start[0]=xxx&and[0]=xxx&id[1]=urn:yyy&passageId[1]=yyy&id[2]=urn:zzz` ???

**Comment by architexte** [comment]
Do we really need multiple sections request URL?
The load balancing is better with multiple requests and in case of time out or error response we can reload in a more efficient way.
The only advantage I see at that moment with multiple passages URL is citability.

**Comment by awagner-mainz** [comment]
I still have to think more thoroughly about it, but something occurring to me immediately is this: from a logical point of view, requesting *multiple* sections is actually requesting *one* (readable) collection, no? I am not clear what the implications of this are; maybe you would pass a parameter to "expand" the collection's members in-place or something like that.

**Comment by PonteIneptique** [comment]
This was dropped in favor of saying to people they should use async http queries.
